### PR TITLE
Migrate update framework to Velopack

### LINKED ISSUE #10550: Wrong GPU selected after osu!lazer update
**Describe the bug:**
I need to manually add osu to the list of programs nvidia will detect in order to run osu on my dedicated gpu (I'm using an old laptop from around 2009).
After updating osu the folder where osu is located has changed because it has the version numbers in it.
Not that it's a big issue but now I have to add the new version to nvidia's list after every update :smile:

**Screenshots or videos showing encountered issue:** 
image

**osu!lazer version:** 
2020.1017.0

### LINKED ISSUE #10588: Taskbar icon not working
**Describe the bug:** 
When osu!lazer is on pinned on windows taskbar and running, the actual program icon is next to it, and it doesn't show the icon.

**Screenshots or videos showing encountered issue:** 
icon

**osu!lazer version:** 
2020.1020.0

**Logs:**

[database.log](https://github.com/ppy/osu/files/5432785/database.log)
[network.log](https://github.com/ppy/osu/files/5432786/network.log)
[performance.log](https://github.com/ppy/osu/files/5432787/performance.log)
[performance-audio.log](https://github.com/ppy/osu/files/5432788/performance-audio.log)
[performance-draw.log](https://github.com/ppy/osu/files/5432789/performance-draw.log)
[performance-update.log](https://github.com/ppy/osu/files/5432790/performance-update.log)
[runtime.log](https://github.com/ppy/osu/files/5432791/runtime.log)
[updater.log](https://github.com/ppy/osu/files/5432792/updater.log)

### LINKED ISSUE #18318: Closing the game during update at just the right time will not open again
### Type

Other

### Bug description

If I close the game during an update at just the right time (discovered by accident), then clicking the shortcut icon on the desktop will not launch the game anymore. Checked in Process Explorer, the executable closes so fast that it doesn't even show up most of the time. Trying to run `%LOCALAPPDATA%\osulazer\osu!.exe` indeed just closes. Running the previous version's exe directly works just fine however.

Seems like it's an issue with the Squirrel stub itself, as it fails even if the new version's folder just exists, empty or not. Deleting the new version's folder "fixes" the issue, and allows the old version to open again, so it can download the new version.

### Screenshots or videos

https://www.youtube.com/watch?v=pLpDbaZWCA4

### Version

2022.509.0

### Logs

[runtime.log](https://github.com/ppy/osu/files/8712102/runtime.log)
[updater.log](https://github.com/ppy/osu/files/8712103/updater.log)
[performance.log](https://github.com/ppy/osu/files/8712104/performance.log)
[database.log](https://github.com/ppy/osu/files/8712105/database.log)
[performance-draw.log](https://github.com/ppy/osu/files/8712106/performance-draw.log)
[performance-update.log](https://github.com/ppy/osu/files/8712107/performance-update.log)

### LINKED ISSUE #23997: Nupkg file blocking updates
### Type

Other

### Bug description

If the update was not completed last time, the nupkg file left in C:\Users\xxx\AppData\Local\osulazer\packages may cause the next update to fail, as mentioned in the video.

### Screenshots or videos

https://github.com/ppy/osu/assets/87159802/8cef922f-8d1d-41b3-b594-2170dbe86b49

### Version

2023.617.0-lazer

### Logs

[runtime.log](https://github.com/ppy/osu/files/11828705/runtime.log)

### PR DESCRIPTION
**Pull request by smallketchup82**
- Should be deployed with https://github.com/ppy/osu-deploy/pull/171
- Closes #18318
- Closes #24484
- Closes #23997
- Closes #10550
- Closes #10588

## Changes
- Refactors & Renames `SimpleUpdateManager` to `MobileUpdateNotifier`
- Removes `SquirrelUpdateManager`
- Implements `VelopackUpdateManager`
- Refactors internal restart logic to utilize Velopack
- Completely removes Clowd.Squirrel as a dependency
- Removes install text from update notifications
	- Velopack won't install the updates while the program is open, it'll do
it in between restarts or before starting.  Therefore the only notification text we need is `StartDownload`, `FailDownload`, etc

## TODO
- [x] Wait for upstream changes (https://github.com/velopack/velopack/issues/151) to get merged & deployed
- [x] Remove SimpleUpdateManager
- [x] Refactor UpdateManager to work with Velo
- [x] Allow downgrading
- [x] Figure out what to do with https://github.com/ppy/osu/blob/9e01cf7fc23812c4d0a76b09dd78ce6b38c06028/osu.iOS/OsuGameIOS.cs#L18 and https://github.com/ppy/osu/blob/f8a2b0724d77721d4fe30fe9a4d4b457efcc73e7/osu.Android/OsuGameAndroid.cs#L83

## Next up

- Adjust windows associations to work with the static executable path (do this in a separate PR)
- Figure out what to do with:
	- https://github.com/ppy/osu/pull/26558
	- https://github.com/ppy/osu/issues/23635
	- https://github.com/ppy/osu/issues/12324

---

NOTE: I haven't tested the update logic because of the difficulty of doing that, and the fact that bugs here would really only show themselves in a production release environment. I have tested the restarting logic and it works.

Read https://github.com/ppy/osu-deploy/pull/171 for information on migration to Velopack

### CODE REVIEW

**Comment by bdach** [comment]
What are the todos in the OP supposed to mean? Is this ready for review or not?

**Comment by smallketchup82** [comment]
> What are the todos in the OP supposed to mean? Is this ready for review or not?

It's ready for review. The todo's remaining are mostly for future reference (things I'll want to add in future PR's).

But the last todo in specific will need a bit of discussion. I asked a question about it earlier in the osu!dev discord:
> though, i did have a small concern (mainly for the core team though). github ended up catching and notifying me that [this](https://github.com/ppy/osu/blob/9e01cf7fc23812c4d0a76b09dd78ce6b38c06028/osu.iOS/OsuGameIOS.cs#L18) exists, and that poses a problem. in my PR i remove SimpleUpdateManager, but i didn't really take into account the fact that mobile builds would be using it, as i assumed they'd use apple's update ecosystem.
> 
> now i'm unsure of whether i should bring SimpleUpdateManager back to cover that singular iOS use case, or whether i just completely remove references to SimpleUpdateManager in osu.iOS

Apart from that concern everything is ready and good to go. Do remember that this needs to be merged at the same time as the osu-deploy PR, as they're codependent.

**Comment by smallketchup82** [comment]
Addressed the last concern ([see for conversation](https://discord.com/channels/188630481301012481/188630652340404224/1258682364817576036)). Should be good for review now.

**Review by smoogipoo — DISMISSED** [review]
Lgtm! Can be tested using releases at https://github.com/smoogipoo/osu/releases.

**Comment by smallketchup82** [comment]
@smoogipoo The velopack version on this has not been updated to reflect https://github.com/ppy/osu-deploy/commit/682917287cdc24b4340c6bf35033e38a702f5b05

I'll update it myself later today or tomorrow, but if you want to quickly do it now feel free.

**Comment by smoogipoo** [comment]
It doesn't really matter, but feel free.

**Comment by smallketchup82** [comment]
Bumped the version to match osu-deploy. You might want to do another quick test & reapprove if everything looks fine. Though, it should be fine given it builds.

**Comment by smoogipoo** [comment]
Have updated the releases in https://github.com/smoogipoo/osu/releases, all looks good.

**Review by smoogipoo — DISMISSED** [review]

File: osu.Desktop/Program.cs [hunk 1]
```diff
@@ -31,19 +30,9 @@ public static class Program
         [STAThread]
         public static void Main(string[] args)
         {
-            /*
```
[the remark this answers is not in the document]
Comment (smallketchup82) [reply osu.Desktop/Program.cs:34]: The Velopack docs suggest that you [*must*](https://arc.net/l/quote/urfqwedh) run it before everything else. However, as you found, that's not necessarily a strict requirement. In another example on Velopack's docs, code is [executed](https://arc.net/l/quote/chgyexjh) before the `.Run()`

As for me, I decided "So you should put `setupVelopack()` before everything else, but putting stuff before it isn't the end of the world as it would've been with Squirrel. So I'll tone it down a bit, and just keep the advisory."

Though, @caesay would likely know best about this. I'd wait for his thoughts.

File: osu.Desktop/OsuGameDesktop.cs [hunk 2]
```diff
@@ -102,35 +100,21 @@ protected override UpdateManager CreateUpdateManager()
...
-                    Debug.Assert(OperatingSystem.IsWindows());
-
-                    // Of note, this is an async method in squirrel that adds an arbitrary delay before returning
-                    // likely to ensure the external process is in a good state.
-                    //
-                    // We're not waiting on that here, but the outro playing before the actual exit should be enough
-                    // to cover this.
-                    Squirrel.UpdateManager.RestartAppWhenExited().FireAndForget();
-                    return true;
+                Velopack.UpdateExe.Start();
```
Comment (peppy) [inline osu.Desktop/OsuGameDesktop.cs:110]: This call is now blocking and will cause a stutter in the game.

We can avoid this with an assumption that the outro sequence will take longer than the updater process takes to start up:

```diff
diff --git a/osu.Desktop/OsuGameDesktop.cs b/osu.Desktop/OsuGameDesktop.cs
index 94f6ef0fc3..c75a3f0a1a 100644
--- a/osu.Desktop/OsuGameDesktop.cs
+++ b/osu.Desktop/OsuGameDesktop.cs
@@ -5,6 +5,7 @@
 using System.IO;
 using System.Reflection;
 using System.Runtime.Versioning;
+using System.Threading.Tasks;
 using Microsoft.Win32;
 using osu.Desktop.Performance;
 using osu.Desktop.Security;
@@ -18,6 +19,7 @@
 using osu.Framework.Allocation;
 using osu.Game.IO;
 using osu.Game.IPC;
+using osu.Game.Online.Multiplayer;
 using osu.Game.Performance;
 using osu.Game.Utils;
 
@@ -105,16 +107,8 @@ protected override UpdateManager CreateUpdateManager()
 
         public override bool RestartAppWhenExited()
         {
-            try
-            {
-                Velopack.UpdateExe.Start();
-                return true;
-            }
-            catch (Exception e)
-            {
-                Logger.Error(e, "Failed to restart application");
-                return base.RestartAppWhenExited();
-            }
+            Task.Run(() => Velopack.UpdateExe.Start()).FireAndForget();
+            return true;
         }
 
         protected override void LoadComplete()

```

Or if we actually need to wait for the process... things are going to get quite a bit more complicated.

@smoogipoo thoughts on this? I think I'd just propose the above to match what we were already doing.

File: osu.Desktop/OsuGameDesktop.cs [hunk 3]
```diff
@@ -102,35 +100,21 @@ protected override UpdateManager CreateUpdateManager()
...
+                return true;
+            }
+            catch (Exception e)
+            {
+                Logger.Error(e, "Failed to restart application");
```
Comment (peppy) [inline osu.Desktop/OsuGameDesktop.cs:115]: This messaging is wrong. It's not failing to restart anything; it's failing to start the updater.

File: osu.Desktop/OsuGameDesktop.cs (hunk 2 above)
Comment (smallketchup82) [reply osu.Desktop/OsuGameDesktop.cs:110]: The above solution works fine. In my [discussion](https://discord.com/channels/188630481301012481/188630652340404224/1255899633687859270) with caesay, he implemented a restart implementation into velopack which should wait 30 seconds for the assembly to exit on its own, otherwise it will kill it. We don't necessarily need to assume anything here, the updater process will wait for us.

Though, I'm a little iffy about doing `FireAndForget`. I would still like to handle exceptions in starting the restarter. Best not to leave the caller of the function in a state of confusion as to whether the game will actually restart or not.

File: osu.Desktop/OsuGameDesktop.cs (hunk 3 above)
Comment (smallketchup82) [reply osu.Desktop/OsuGameDesktop.cs:115]: > it's failing to start the updater.

Which, in turn, results in a failure to restart the application. I can make it more specific, but it's not entirely wrong to say that restarting the application failed (since that *is* the purpose of the method). Would you rather I completely reword it to "Failed to launch restarter process", or clarify with "Failed to start restarter process. A restart will not be preformed"

File: osu.Desktop/OsuGameDesktop.cs (hunk 2 above)
Comment (smoogipoo) [reply osu.Desktop/OsuGameDesktop.cs:110]: I think it's fine to match what we were already doing.

File: osu.Desktop/OsuGameDesktop.cs (hunk 3 above)
Comment (peppy) [reply osu.Desktop/OsuGameDesktop.cs:115]: Don't change anything; see the other comment above which removes this completely.

File: osu.Desktop/OsuGameDesktop.cs (hunk 2 above)
Comment (peppy) [reply osu.Desktop/OsuGameDesktop.cs:110]: > No need to worry about pacing/race conditions here.

The fail case would the update process *not being started* in time, not the game *not being killed in time*. It's a different concern, but as I mentioned it's very unlikely to cause a problem. Worst case the user has to start osu! again because it didn't do so automatically.

File: osu.Desktop/Program.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Desktop/Program.cs:34]: I've made the inline comment a bit louder for now. Probably fine to continue.

**Review by peppy — DISMISSED** [review]

**Comment by caesay** [comment]
> The velopack version on this has not been updated to reflect https://github.com/ppy/osu-deploy/commit/682917287cdc24b4340c6bf35033e38a702f5b05 
> It doesn't really matter, but feel free.

Just a warning that sometimes things do break if you are using different versions of the Velopack NuGet and VPK. I try to minimize this of course, but as an example I changed the default behavior of Update.exe recently from "don't restart by default, require --restart to turn on" to "restart by default and turn off with --norestart". The if you were using a mismatched package version at that time, your app may not/restart as you expect. The VPK tool will emit a warning if the version is not the same but you might not catch that in CI. Just keep that in mind!

File: osu.Desktop/Updater/VelopackUpdateManager.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                    Scheduler.AddDelayed(() => Task.Run(async () => await checkForUpdateAsync().ConfigureAwait(false)), 60000 * 30);
+                }
+            }
+
+            return true;
+        }
+
+        private bool restartToApplyUpdate()
+        {
+            updateManager.WaitExitThenApplyUpdates(null);
```
Comment (caesay) [inline osu.Desktop/Updater/VelopackUpdateManager.cs:122]: Really need to provide an UpdateInfo here instead of null, the same one you passed to `DownloadUpdatesAsync`.

**Comment by peppy** [comment]
The package version has since been updated in line with `osu-deploy` FWIW.

**Comment by caesay** [comment]
This should close these issues also #18318 #24484 #23997 #10550

**Comment by peppy** [comment]
@smallketchup82 please don't resolve review threads, we leave these unresolved if they are still applicable (and should be read by other reviewers).

File: osu.Desktop/Updater/VelopackUpdateManager.cs (hunk 4 above)
Comment (peppy) [reply osu.Desktop/Updater/VelopackUpdateManager.cs:122]: @caesay how should we handle the case where `UpdateInfo` came back null but there's a pending update?

JetBrains Rider 2024-09-02 at 08 14 41

File: osu.Desktop/OsuGameDesktop.cs (hunk 2 above)
Comment (caesay) [reply osu.Desktop/OsuGameDesktop.cs]: There is nothing blocking about UpdateExe.Start. there does appear to be a Thread.Sleep(300) in UpdateExe.Apply but I could refactor this to a Task.Delay

File: osu.Desktop/Updater/VelopackUpdateManager.cs (hunk 4 above)
Comment (peppy) [reply osu.Desktop/Updater/VelopackUpdateManager.cs:122]: Based on the xmldoc it seems like passing null is fine in this case "The target release to apply. Can be left null to auto-apply the newest downloaded release."

Thanks for good documentation!

Comment (peppy) [reply osu.Desktop/Updater/VelopackUpdateManager.cs]: Applied in https://github.com/ppy/osu/pull/28743/commits/cd9b82253e4639d3a94dc049244c9ccbcc59ea47.

Comment (caesay) [reply osu.Desktop/Updater/VelopackUpdateManager.cs]: Maybe I should refactor `IsUpdatePendingRestart` to return a `VelopackAsset` so it can be passed in. Null will probably work most of the time but it's not best practice.

If you pass null, Velopack doesn't know for sure what update you actually want to apply, so it will just go searching in the packages directory for the latest version. Sometimes this is not what you want. For example, Velopack supports updating to older versions now - both explicitly and implicitly eg. it can automatically downgrade if you've retracted a release from your feed because it has breaking bugs as long as the `UpdateOptions.AllowVersionDowngrade` is set to true. 

So if you are downgrading a version but have not provided it to the apply/restart then you end up not installing it.

Also, DownloadUpdates should be a no-op if the release is already downloaded. 

My suggestion is you should not need to check IsUpdatePendingRestart except for on startup (but VelopackApp already does this for you by default.. see SetAutoApply there). 

You can just CheckForUpdates, if it's not null, call DownloadUpdates and then cache the UpdateInfo in a local field. If you have the local field non-null when a restart is requested then you know what to pass to Velopack.. If your local UpdateInfo is null, then you've not downloaded a new version yet.

Comment (peppy) [reply osu.Desktop/Updater/VelopackUpdateManager.cs]: > see SetAutoApply

Hmm, this may be something we want to turn off to avoid the game hanging with no visible progress on startup for 10 seconds or more, I guess? This is going to block startup while the diff is applied correct?

File: osu.Desktop/OsuGameDesktop.cs (hunk 2 above)
Comment (peppy) [reply osu.Desktop/OsuGameDesktop.cs]: Hmm, I have noticed a pause on macOS. I've have to profile to be sure, but I think `Process.Start` can block for short time.

File: osu.Desktop/Updater/VelopackUpdateManager.cs (hunk 4 above)
Comment (caesay) [reply osu.Desktop/Updater/VelopackUpdateManager.cs]: It does restart and apply on startup yes. On windows there is a progress dialog, nothing for mac/linux but I'm working on that. I recently just finished a cross-platform dialog library https://github.com/velopack/xdialog-rs but it has not yet been integrated to Velopack.

Comment (caesay) [reply osu.Desktop/Updater/VelopackUpdateManager.cs]: On a faster PC this should only take a second or two, since it's just unzipping the nupkg. On a slower PC it's a bit unfortunate without the progress bar there.

Comment (peppy) [reply osu.Desktop/Updater/VelopackUpdateManager.cs]: > You can just CheckForUpdates, if returns not null, call DownloadUpdates and then cache the UpdateInfo in a local field. If you have the local field non-null when a restart is requested then you know what to pass to Velopack.. If your local UpdateInfo is null, then you've not downloaded a new version yet.

Could there be a case where a second check-for-updates finds a newer update, where we'd want to use the newer `UpdateInfo`? Or is that not something that can happen in the update flow?

File: osu.Desktop/OsuGameDesktop.cs (hunk 2 above)
Comment (peppy) [reply osu.Desktop/OsuGameDesktop.cs]: @caesay profiled this. as you say, it's mostly the sleep:

JetBrains Rider 2024-09-04 at 06 50 17

Of note, the overhead of starting the process *may still cause a frame stutter for users with higher refresh rates*. But the sleep part is the bigger issue.

Making this flow async would probably be the best outcome.

**Review by peppy — APPROVED** [review]

File: osu.Desktop/Updater/VelopackUpdateManager.cs (hunk 4 above)
Comment (caesay) [reply osu.Desktop/Updater/VelopackUpdateManager.cs]: It depends on how you implement things. CheckForUpdates will always return the latest update available. If you have already downloaded an update and another release has happened you can do another DownloadUpdates to grab the fresh version again.

Comment (peppy) [reply osu.Desktop/Updater/VelopackUpdateManager.cs]: Okay good to know, thanks.

With my latest changes on this branch I bypassed that, but we'll probably add back support for it for fun.
